python/mxnet/symbol.py

```diff
@@ -508,32 +592,67 @@ def simple_bind(self, ctx, grad_req='write', **kwargs):
             - 'write' means everytime gradient is write to specified args_grad NDArray.
             - 'add' means everytime gradient is add to the specified NDArray.
             - 'null' means no action is taken, the gradient may not be calculated.
-        kwargs : dict of str->shape
+        shape_dict : dict of str->shape
             Input shape dictionary, name->shape
+        type_dict  : dict of str->numpy.dtype
+            Input type dictionary, name->dtype
 
         Returns
         -------
         executor : mxnet.Executor
             The generated Executor
         """
-        arg_shapes, _, aux_shapes = self.infer_shape(**kwargs)
-        if arg_shapes == None:
+        # pylint: disable=too-many-locals
+        arg_shapes, _, aux_shapes = self.infer_shape(**shape_dict)
+        arg_types, _, aux_types = self.infer_type(**type_dict)
+        arg_types = [_DTYPE_MX_TO_NP[i] for i in arg_types]
+        aux_types = [_DTYPE_MX_TO_NP[i] for i in aux_types]
+        if arg_shapes == None or arg_types == None:
             raise ValueError("Input node is not complete")
         # alloc space
-        arg_ndarrays = [zeros(shape, ctx) for shape in arg_shapes]
+        arg_ndarrays = [zeros(shape, ctx, dtype=dtype)for dtype, shape in zip(arg_types,
+                                                                              arg_shapes)]
 
         if grad_req != 'null':
             grad_ndarrays = {}
-            for name, shape in zip(self.list_arguments(), arg_shapes):
+            for name, shape, dtype in zip(self.list_arguments(), arg_shapes, arg_types):
                 if not (name.endswith('data') or name.endswith('label')):
-                    grad_ndarrays[name] = zeros(shape, ctx)
+                    grad_ndarrays[name] = zeros(shape, ctx, dtype=dtype)
         else:
             grad_ndarrays = None
 
-        aux_ndarrays = [zeros(shape, ctx) for shape in aux_shapes]
+        aux_ndarrays = [zeros(shape, ctx, dtype=dtype) for shape, dtype in zip(aux_shapes,
+                                                                               aux_types)]
         executor = self.bind(ctx, arg_ndarrays, grad_ndarrays, grad_req, aux_ndarrays)
         return executor
 
+    def simple_bind(self, ctx, grad_req='write', **kwargs):
```

Comment:
Keep the function as simple_bind, by allowing type_dict to be optional as None and use the default when None is detected